### PR TITLE
fix: close opened connections pointing to a dead node

### PR DESCRIPTION
**Pull request by karenc-bq**
### Summary

fix: close opened connections pointing to a dead node

### CODE REVIEW

File: wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java [hunk 1]
```diff
@@ -57,6 +60,7 @@
 public class FailoverConnectionPlugin extends AbstractConnectionPlugin {
 
   private static final Logger LOGGER = Logger.getLogger(FailoverConnectionPlugin.class.getName());
+  private static final Map<String, Queue<Connection>> openedConnections = new HashMap<>();
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java:63]: Map<String, Queue<WeakReference<Connection>>> ?

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import java.sql.SQLException;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+public class ConnectionInvalidator {
+
+  public static final Map<String, Queue<WeakReference<Connection>>> OPENED_CONNECTION =
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:30]: Usually variables in caps are constants.

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      WeakReference<Connection> connReference;
+      while ((connReference = connectionQueue.poll()) != null) {
+        final Connection conn = connReference.get();
+        if (conn == null) {
+          continue;
+        }
+
+        // Try to call rollback in case the connection is in
+        try {
+          conn.rollback();
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:59]: Hmm I'm not sure we need it. Calling rollback() might trigger failover that takes time. After failover is completed we close connection.

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        } catch (final SQLException e) {
+          // swallow this exception
+        }
+        try {
+          if (!conn.isClosed()) {
+            conn.close();
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:65]: abort() ?

File: wrapper/src/main/java/software/amazon/jdbc/PropertyDefinition.java [hunk 5]
```diff
@@ -61,6 +61,13 @@ public class PropertyDefinition {
       new AwsWrapperProperty(
           "database", null, "Driver database name");
 
+  public static final AwsWrapperProperty INVALIDATE_ALL_CONNECTIONS =
+      new AwsWrapperProperty(
+          "invalidateAllConnections",
+          "true",
+          "Invalidate all other connections in a connection pool to the same node if one "
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/PropertyDefinition.java:68]: connection pool? There's no connection pool managed by driver. I'd rephrase it.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/DefaultConnectionPlugin.java [hunk 6]
```diff
@@ -140,6 +145,13 @@ public Connection connect(
 
     final Connection conn = this.connectionProvider.connect(driverProtocol, hostSpec, props);
 
+    if (PropertyDefinition.INVALIDATE_ALL_CONNECTIONS.getBoolean(props)) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/DefaultConnectionPlugin.java:148]: It should be implemented as a call to PluginService. The only way plugin can communicate to outer world is though PluginService.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java [hunk 7]
```diff
@@ -726,12 +707,23 @@ protected void invalidateCurrentConnection() {
...
+  protected void invalidateCurrentConnection(final Connection conn) {
+    if (conn == null) {
+      return;
+    }
+
+    final HostSpec originalHost = this.pluginService.getCurrentHostSpec();
+    isInTransaction = this.pluginService.isInTransaction();
+
+    this.pluginService.invalidateAllConnections(originalHost);
+    this.pluginService.setAvailability(originalHost.getAliases(), HostAvailability.NOT_AVAILABLE);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java:724]: Debatable. This node maybe available (since failover is completed) but it may have a different role.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java [hunk 8]
```diff
@@ -696,28 +696,9 @@ protected void failoverWriter() throws SQLException {
...
-    if (conn == null) {
-      return;
-    }
+    invalidateCurrentConnection(conn);
 
     final HostSpec originalHost = this.pluginService.getCurrentHostSpec();
-    if (this.pluginService.isInTransaction()) {
-      isInTransaction = this.pluginService.isInTransaction();
-      try {
-        conn.rollback();
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java:707]: Need to keep this logic trying rolling back a transaction.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java [hunk 9]
```diff
@@ -726,12 +707,23 @@ protected void invalidateCurrentConnection() {
...
+    }
+
+    final HostSpec originalHost = this.pluginService.getCurrentHostSpec();
+    isInTransaction = this.pluginService.isInTransaction();
+
+    this.pluginService.invalidateAllConnections(originalHost);
+    this.pluginService.setAvailability(originalHost.getAliases(), HostAvailability.NOT_AVAILABLE);
+  }
+
   protected synchronized void pickNewConnection() throws SQLException {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/failover/FailoverConnectionPlugin.java:727]: When failover is processed, we need to invalidate nodes that changed their roles. I haven't found this logic. Do we need it?

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+          hostSpec.getUrl());
+
+      WeakReference<Connection> connReference;
+      while ((connReference = connectionQueue.poll()) != null) {
+        final Connection conn = connReference.get();
+        if (conn == null) {
+          continue;
+        }
+
+        // Try to call rollback in case the connection is in
```
Comment (davecramer) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:57]: is in what ?

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java (hunk 3 above)
Comment (davecramer) [reply wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:59]: if this is in a pool, but not open in the application then close() would have been called. 
We should be able to figure out if the connection is in the app or in the pool.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+  static final String PG_INSTANCE_NAME_COL = "aurora_db_instance_identifier";
+  static final String METHOD_ABORT = "abort";
+  static final String METHOD_CLOSE = "close";
+  private static final String PG_DRIVER_PROTOCOL = "postgresql";
+  private static final Set<String> subscribedMethods =
+      Collections.unmodifiableSet(new HashSet<String>() {
+        {
+          add("connect");
+          add("close");
+          add("notifyNodeListChanged");
```
Comment (karenc-bq) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:54]:
```suggestion
          add("notifyNodeListChanged");
          add("initHostListProvider");
```

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+            final Thread abortThread = new Thread(r);
+            abortThread.setDaemon(true);
+            return abortThread;
+          });
+
+  private static final Logger LOGGER = Logger.getLogger(ConnectionInvalidator.class.getName());
+  private static final RdsUtils rdsUtils = new RdsUtils();
+
+  public static void populateOpenedConnectionQueue(final HostSpec hostSpec, final Connection conn) {
+    final Set<String> aliases = hostSpec.getAliases().isEmpty()
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:59]: You can simply use hostSpec.asAlias()

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        ? new HashSet<>(Collections.singletonList(hostSpec.asAlias()))
+        : hostSpec.getAliases();
+
+    // Check if the connection was established using a cluster endpoint
+    final RdsUrlType urlType = rdsUtils.identifyRdsType(hostSpec.getHost());
+    String endpoint = null;
+    if (urlType.isRdsCluster()) {
+      LOGGER.info("Attempting to populate opened connection queue with cluster endpoint.");
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:67]: LOGGER.finest() ?

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+
+      // 1. Check if aliases contain an instance endpoint.
+      LOGGER.info(String.format("host instance: %s\n", hostSpec.getHost()));
+
+      for (String alias : aliases) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:72]: Full host name is in hostSpec.getHost(). Aliases are IP addresses, instance nemes, etc. I believe this loop can be removed.

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+  private static Queue<WeakReference<Connection>> getConnectionQueue(Set<String> aliases) {
+    Queue<WeakReference<Connection>> connectionQueue;
+
+    for (String alias : aliases) {
+      if (openedConnections.containsKey(alias)) {
+        connectionQueue = openedConnections.get(alias);
+        if (connectionQueue != null) {
+          return connectionQueue;
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:125]: Should we continue with loop? Other aliases may give other queues of connections.

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionPluginManager.java [hunk 16]
```diff
@@ -70,6 +71,7 @@ public class ConnectionPluginManager implements CanReleaseResources {
           put("awsSecretsManager", AwsSecretsManagerConnectionPluginFactory.class);
           put("auroraStaleDns", AuroraStaleDnsPluginFactory.class);
           put("readWriteSplitting", ReadWriteSplittingPluginFactory.class);
+          put("connectionDestroyer", AuroraConnectionDestroyerPluginFactory.class);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionPluginManager.java:74]: We need to come up to a better name, less like "death star". :)

Comment (sergiyvamz) [reply wrapper/src/main/java/software/amazon/jdbc/ConnectionPluginManager.java:74]: It should start with "aurora"

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+      final String host = currentHostSpec.getHost();
+      if (rdsHelper.isRdsClusterDns(currentHostSpec.getHost())) {
+        currentHostSpec.addAlias(getInstanceEndpoint(conn, host));
+      }
+    }
+
+    ConnectionInvalidator.populateOpenedConnectionQueue(hostSpec, conn);
+    ConnectionInvalidator.logOpenedConnections();
+
+    return super.connect(driverProtocol, hostSpec, props, isInitialConnection, connectFunc);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:100]: return conn;

Comment (sergiyvamz) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:100]: If you want to use super.connect() it should be done at line 87.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  @Override
+  public void initHostProvider(String driverProtocol, String initialUrl, Properties props,
+      HostListProviderService hostListProviderService,
+      JdbcCallable<Void, SQLException> initHostProviderFunc) throws SQLException {
+    if (hostListProviderService.isStaticHostListProvider()) {
+      hostListProviderService.setHostListProvider(new AuroraHostListProvider(
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:108]: Why do we need Aurora topology provider in this plugin?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+  public <T, E extends Exception> T execute(Class<T> resultClass, Class<E> exceptionClass,
+      Object methodInvokeOn, String methodName, JdbcCallable<T, E> jdbcMethodFunc,
+      Object[] jdbcMethodArgs) throws E {
+
+    final HostSpec originalHost = this.pluginService.getCurrentHostSpec();
+
+    try {
+      T result = jdbcMethodFunc.call();
+      if ((methodName.contains(METHOD_CLOSE) || methodName.contains(METHOD_ABORT))) {
+        ConnectionInvalidator.invalidateAllConnections(originalHost);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:136]: You probably need to remove the current connection from all queues, no to close all opened connections to the same node.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+      }
+      return result;
+    } catch (Exception e) {
+      if (e instanceof FailoverSQLException) {
+        ConnectionInvalidator.invalidateAllConnections(originalHost);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:141]: Do we need to check what node changed their roles?

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.logging.Logger;
+import software.amazon.jdbc.util.RdsUrlType;
+import software.amazon.jdbc.util.RdsUtils;
+
+public class ConnectionInvalidator {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:36]: OpenConnectionTracker ?

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java (hunk 13 above)
Comment (karenc-bq) [reply wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:67]: These are debug logs, I think they are best taken out of the final code since they don't really contribute much.

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java (hunk 14 above)
Comment (karenc-bq) [reply wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java:72]: Not sure if I follow. We only get to this loop if hostSpec.getHost() returns a cluster endpoint, which is not what we want.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java (hunk 19 above)
Comment (karenc-bq) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:136]: That makes sense

File: wrapper/src/main/java/software/amazon/jdbc/ConnectionPluginManager.java [hunk 22]
```diff
@@ -70,6 +71,7 @@ public class ConnectionPluginManager implements CanReleaseResources {
           put("awsSecretsManager", AwsSecretsManagerConnectionPluginFactory.class);
           put("auroraStaleDns", AuroraStaleDnsPluginFactory.class);
           put("readWriteSplitting", ReadWriteSplittingPluginFactory.class);
+          put("auroraConnectionTracker", AuroraConnectionDestroyerPluginFactory.class);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/ConnectionPluginManager.java:74]: Let's change name of the factory class: AuroraConnectionDestroyerPluginFactory

File: wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.logging.Logger;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.RdsUrlType;
+import software.amazon.jdbc.util.RdsUtils;
+
+public class OpenedConnectionTracker {
+
+  public static final Map<String, Queue<WeakReference<Connection>>> openedConnections = new ConcurrentHashMap<>();
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java:38]: It shouldn't be public

File: wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+
+  public static void populateOpenedConnectionQueue(final HostSpec hostSpec, final Connection conn) {
+    final Set<String> aliases = hostSpec.asAliases();
+
+    // Check if the connection was established using a cluster endpoint
+    final RdsUrlType urlType = rdsUtils.identifyRdsType(hostSpec.getHost());
+    String endpoint = null;
+    if (urlType.isRdsCluster()) {
+      // 1. Check if aliases contain an instance endpoint.
+      for (String alias : aliases) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java:65]: Can we do just  simple check:
aliases.contains(hostSpec.getHost()) 
?

File: wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+        }
+      }
+
+      if (endpoint == null) {
+        LOGGER.fine(Messages.get("OpenedConnectionTracker.unableToPopulateOpenedConnectionQueue"));
+        return;
+      }
+    }
+
+    Queue<WeakReference<Connection>> connectionQueue = getConnectionQueue(aliases);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java:78]: getConnectionQueue(endpoint) ?

File: wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+    connectionQueue.add(new WeakReference<>(conn));
+
+    for (String alias : aliases) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java:81]: It seems we don't need this code since connection should always be in the map wit instance endpoint as a key.

File: wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+      if (rdsUtils.isRdsInstance(alias)) {
+        openedConnections.put(alias, connectionQueue);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java:83]: getConnectionQueue() can return an existing queue or create a new one.

File: wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+   * @param hostSpec The {@link HostSpec} object containing the url of the node.
+   */
+  public static void invalidateAllConnections(final HostSpec hostSpec) {
+    invalidateAllConnections(hostSpec.getAliases().toArray(new String[] {}));
+  }
+
+  public static void invalidateAllConnections(final String... node) {
+    Queue<WeakReference<Connection>> connectionQueue = getConnectionQueue(new HashSet<>(Arrays.asList(node)));
+
+    if (connectionQueue.isEmpty()) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java:102]: invalidateConnections() can handle this

File: wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+
+  public static void invalidateCurrentConnection(final HostSpec hostSpec, final Connection connection) {
+    Queue<WeakReference<Connection>> connectionQueue = getConnectionQueue(hostSpec.getAliases());
+    // connectionQueue.removeIf(connectionWeakReference -> Objects.equals(connectionWeakReference.get(), connection));
+  }
+
+  private static Queue<WeakReference<Connection>> getConnectionQueue(Set<String> aliases) {
+    Queue<WeakReference<Connection>> connectionQueue;
+
+    for (String alias : aliases) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java:117]: Don't need to check aliases as agreed before.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 30]
```diff
@@ -0,0 +1,161 @@
...
+import software.amazon.jdbc.NodeChangeOptions;
+import software.amazon.jdbc.OpenedConnectionTracker;
+import software.amazon.jdbc.PluginService;
+import software.amazon.jdbc.hostlistprovider.AuroraHostListProvider;
+import software.amazon.jdbc.plugin.failover.FailoverSQLException;
+import software.amazon.jdbc.util.RdsUtils;
+import software.amazon.jdbc.util.StringUtils;
+import software.amazon.jdbc.util.SubscribedMethodHelper;
+
+public class AuroraConnectionDestroyerPlugin extends AbstractConnectionPlugin {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:41]: AuroraConnectionTrackerPlugin ?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 31]
```diff
@@ -0,0 +1,161 @@
...
+  static final String MYSQL_GET_INSTANCE_NAME_COL = "@@aurora_server_id";
+  static final String PG_GET_INSTANCE_NAME_SQL = "SELECT aurora_db_instance_identifier()";
+  static final String PG_INSTANCE_NAME_COL = "aurora_db_instance_identifier";
+  static final String METHOD_ABORT = "abort";
+  static final String METHOD_CLOSE = "close";
+  private static final String PG_DRIVER_PROTOCOL = "postgresql";
+  private static final Set<String> subscribedMethods =
+      Collections.unmodifiableSet(new HashSet<String>() {
+        {
+          addAll(SubscribedMethodHelper.NETWORK_BOUND_METHODS);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:53]: Why?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 32]
```diff
@@ -0,0 +1,161 @@
...
+    final Connection conn = connectFunc.call();
+
+    if (conn != null) {
+      final HostSpec currentHostSpec = this.pluginService.getCurrentHostSpec();
+      if (rdsHelper.isRdsClusterDns(currentHostSpec.getHost())) {
+        currentHostSpec.addAlias(getInstanceEndpoint(conn, currentHostSpec));
+      }
+    }
+
+    OpenedConnectionTracker.populateOpenedConnectionQueue(hostSpec, conn);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:97]: Can we make populateOpenedConnectionQueue() not static? It'd allow us to pass a reference of OpenedConnectionTracker as a plugin dependency and simplify unit testing. Static dependencies are hard to mock.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 33]
```diff
@@ -0,0 +1,161 @@
...
+      OpenedConnectionTracker.logOpenedConnections();
+      if (e instanceof FailoverSQLException) {
+        OpenedConnectionTracker.invalidateAllConnections(originalHost);
+      }
+      throw e;
+    }
+  }
+
+  @Override
+  public void notifyNodeListChanged(Map<String, EnumSet<NodeChangeOptions>> changes) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:134]: During failover this notification may be quite noisy and gives false positive. Can we reply just on failover exceptions instead?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 34]
```diff
@@ -0,0 +1,161 @@
...
+import software.amazon.jdbc.util.StringUtils;
+import software.amazon.jdbc.util.SubscribedMethodHelper;
+
+public class AuroraConnectionDestroyerPlugin extends AbstractConnectionPlugin {
+
+  static final String MYSQL_GET_INSTANCE_NAME_SQL = "SELECT @@aurora_server_id";
+  static final String MYSQL_GET_INSTANCE_NAME_COL = "@@aurora_server_id";
+  static final String PG_GET_INSTANCE_NAME_SQL = "SELECT aurora_db_instance_identifier()";
+  static final String PG_INSTANCE_NAME_COL = "aurora_db_instance_identifier";
+  static final String METHOD_ABORT = "abort";
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:47]: "Connection.abort"

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java [hunk 35]
```diff
@@ -0,0 +1,161 @@
...
+  static final String METHOD_CLOSE = "close";
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPlugin.java:48]: "Connection.close"

File: wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPluginFactory.java [hunk 36]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package software.amazon.jdbc.plugin;
+
+import java.util.Properties;
+import software.amazon.jdbc.ConnectionPlugin;
+import software.amazon.jdbc.ConnectionPluginFactory;
+import software.amazon.jdbc.PluginService;
+
+public class AuroraConnectionDestroyerPluginFactory implements ConnectionPluginFactory {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/AuroraConnectionDestroyerPluginFactory.java:24]: AuroraConnectionTrackerPluginFactory

File: wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java [hunk 37]
```diff
@@ -110,4 +119,69 @@ public void testOpenConnectionWithDataSourceClassName() throws SQLException {
...
+            .getClusterEndpoint());
+    targetDataSourceProps.setProperty(
+        "databaseName",
+        TestEnvironment.getCurrent().getInfo().getDatabaseInfo().getDefaultDbName());
+    targetDataSourceProps.setProperty("url", ConnectionStringHelper.getClusterUrl());
+
+    targetDataSourceProps.setProperty(PropertyDefinition.PLUGINS.name, "auroraConnectionTracker,failover");
+    ds.addDataSourceProperty("targetDataSourceProperties", targetDataSourceProps);
+
+    auroraUtil.queryInstanceId(
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java:159]: Do we need results of this call?

File: wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java [hunk 38]
```diff
@@ -110,4 +119,69 @@ public void testOpenConnectionWithDataSourceClassName() throws SQLException {
...
+            () ->
+                auroraUtil.queryInstanceId(
+                    TestEnvironment.getCurrent().getInfo().getRequest().getDatabaseEngine(),
+                    ds.getConnection()));
+    assertEquals("08S02", exception.getSQLState());
+
+    // Ensure that all idle connections are closed.
+    // HikariCP will throw a SQLTransientConnectionException if the connection is closed or invalid.
+    for (int i = 0; i < POOL_SIZE; i++) {
+      assertThrows(SQLTransientConnectionException.class, () -> ds.getConnection().isValid(3));
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java:184]: Hikari can open new connection when calling ds.getConnection() and that may ruin the purpose of the test. We need to come up to another solution to check connections that are already in the pool and make sure there's no valid (i.e. not yet closed) connections to nodes from line 159 and line 170.

Comment (sergiyvamz) [reply wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java:184]: Should we check isClosed() instead?

File: wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java [hunk 39]
```diff
@@ -110,4 +119,69 @@ public void testOpenConnectionWithDataSourceClassName() throws SQLException {
...
+        TestEnvironment.getCurrent().getInfo().getDatabaseInfo().getDefaultDbName());
+    targetDataSourceProps.setProperty("url", ConnectionStringHelper.getClusterUrl());
+
+    targetDataSourceProps.setProperty(PropertyDefinition.PLUGINS.name, "auroraConnectionTracker,failover");
+    ds.addDataSourceProperty("targetDataSourceProperties", targetDataSourceProps);
+
+    auroraUtil.queryInstanceId(
+        TestEnvironment.getCurrent().getInfo().getRequest().getDatabaseEngine(),
+        ds.getConnection());
+
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java:162]: Does it make sense to check how many connections are we currently tracking?

**Review by sergiyvamz — APPROVED** [review]

File: wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java (hunk 37 above)
Comment (karenc-bq) [reply wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java:171]: No

File: wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java (hunk 38 above)
Comment (karenc-bq) [reply wrapper/src/test/java/integration/refactored/container/tests/HikariTests.java]: Both isClosed and isValid returns the same result. The exception is thrown by Hikari before any database execution

**Comment by davecramer** [comment]
Should probably remove Aurora from the name of the plugin. Is it specific to Aurora ?
